### PR TITLE
calloc limits

### PR DESCRIPTION
**Pull request by DellaBitta**
Adds overflow protection to `__iotc_calloc` calls, as well as unit tests for the function.

### CODE REVIEW

**Review by atigyi — COMMENTED** [review]
David, thx for jumping on this topic so quick!! Added few comments. I like the utests!! :)

File: src/tests/utests/iotc_utest_memory_calloc.c [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    {
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/0);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__zero_num__zero_size,
+    {
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/0);
```
[the remark this answers is not in the document]
Comment (DellaBitta) [reply src/tests/utests/iotc_utest_memory_calloc.c:64]: 👍 nice, thanks.

File: src/tests/utests/iotc_utest_memory_calloc.c [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      void* ptr = __iotc_calloc( /*num=*/SIZE_MAX/2, /*size=*/2);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__overflow__num_two__half_max_size,
+    {
+      void* ptr = __iotc_calloc( /*num=*/2, /*size=*/SIZE_MAX/2);
+      tt_want_ptr_op(NULL, ==, ptr);
```
[the remark this answers is not in the document]
Comment (DellaBitta) [reply src/tests/utests/iotc_utest_memory_calloc.c:103]: Yeah, interesting, right?  Unfortunately behavior seems to vary across platforms.  Allocating a SIZE_MAX produces `vApplicationMallocFailedHook` errors on AMD64 travis runners (see [Job #606.5 errored](https://travis-ci.com/github/GoogleCloudPlatform/iot-device-sdk-embedded-c/jobs/496593558)) but passes on other runners and in local testing.

I'm working to determine the best cross section of passing tests and leaving comments within the test file for those scenarios I've attempted but which produce inconsistent results.  It seems that it's mainly the behavior around creating allocations of SIZE_MAX, but I'll also add SIZE_MAX -1 tests.

File: src/tests/utests/iotc_utest_memory_calloc.c [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+#endif  // IOTC_TT_TESTCASE_ENUMERATION__SECONDPREPROCESSORRUN
+
+IOTC_TT_TESTGROUP_BEGIN(utest_memory_calloc)
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc,
+    {
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/8);
+      tt_want_ptr_op(NULL, !=, ptr);
```
[the remark this answers is not in the document]
Comment (DellaBitta) [reply src/tests/utests/iotc_utest_memory_calloc.c:41]: sizeof(*(void*)) is a constant - one byte in my testing on Mac.  I don't believe there's a standard check in C to determine an allocation's length.

[the remark this answers is not in the document]
Comment (atigyi) [reply src/tests/utests/iotc_utest_memory_calloc.c:42]: [Seems there isn't](https://stackoverflow.com/questions/1281686/determine-size-of-dynamically-allocated-memory-in-c), unfortunately.

File: src/tests/utests/iotc_utest_memory_calloc.c [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    {
+      void* ptr = __iotc_calloc( /*num=*/0, /*size=*/SIZE_MAX);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+/* Edge cases - SIZE_MAX
+   Note: FreeRTOS heap size isn't large enough for SIZE_MAX allocations,
+   so these tests are disabled for those targets. */
+#ifndef IOTC_PLATFORM_IS_FREERTOS
```
Comment (atigyi) [inline src/tests/utests/iotc_utest_memory_calloc.c:106]: This should work, doesn't it?

File: src/tests/utests/iotc_utest_memory_calloc.c [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/SIZE_MAX);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_max_minus_1__size_1,
+    {
+      void* ptr = __iotc_calloc( /*num=*/SIZE_MAX-1, /*size=*/1);
+      tt_want_ptr_op(NULL, ==, ptr);
```
Comment (atigyi) [inline src/tests/utests/iotc_utest_memory_calloc.c:119]: Hmm, still cannot allocate `SIZE_MAX - 1`?

File: make/mt-os/mt-freertos-linux.mk [hunk 6]
```diff
@@ -16,7 +16,7 @@
 CC ?= gcc
 AR ?= ar
 
-IOTC_COMMON_COMPILER_FLAGS += -fPIC
+IOTC_COMMON_COMPILER_FLAGS += -fPIC -DIOTC_TARGET_FREERTOS=1
```
Comment (atigyi) [inline make/mt-os/mt-freertos-linux.mk:19]: `IOTC_TARGET_FREERTOS` probably this isn't required since there is already a specific for FreeRTOS: `IOTC_PLATFORM_IS_FREERTOS `

File: src/tests/utests/iotc_utest_memory_calloc.c (hunk 4 above)
Comment (DellaBitta) [reply src/tests/utests/iotc_utest_memory_calloc.c]: No, unfortunately it doesn't though I did try it!  

We compile our FreeRTOS tests on linux and so our make system defines `IOTC_PLATFORM_BASE_POSIX`. The string `IOTC_PLATFORM_IS_FREERTOS` seems to be a makefile flag only -- it is not added to our clfags.

Then I didn't want to reuse the Makefile flag `IOTC_PLATFORM_IS_FREERTOS` in our clfags because that would add `-DIOTC_PLATFORM_IS_FREERTOS`  alongside `-DIOTC_PLATFORM_BASE_POSIX`.  Both would appear on the command line and I thought that nomenclature would be confusing, so I named the flag `IOTC_TARGET_FREERTOS`.

File: src/tests/utests/iotc_utest_memory_calloc.c (hunk 5 above)
Comment (DellaBitta) [reply src/tests/utests/iotc_utest_memory_calloc.c:119]: Nope.   FreeRTOS default heap size is much smaller:

In FreeRTOSConfig.h:
```
#define configTOTAL_HEAP_SIZE ((size_t)(350 * 1024))
```

File: make/mt-os/mt-freertos-linux.mk (hunk 6 above)
Comment (DellaBitta) [reply make/mt-os/mt-freertos-linux.mk:19]: Please see my note above.

File: src/tests/utests/iotc_utest_memory_calloc.c (hunk 5 above)
Comment (DellaBitta) [reply src/tests/utests/iotc_utest_memory_calloc.c:119]: Oh I see what you're aying. I think this errro is less of "what does calloc due for performance" and more about how much heap the allocator can put aside in one contiguous. block.   Seems like it can't do SIZE_MAX-1 either, even on my mac and linux boxes.

File: make/mt-os/mt-freertos-linux.mk (hunk 6 above)
Comment (atigyi) [reply make/mt-os/mt-freertos-linux.mk:19]: True, ok, thx for the explanation.

File: src/tests/utests/iotc_utest_memory_calloc.c (hunk 5 above)
Comment (atigyi) [reply src/tests/utests/iotc_utest_memory_calloc.c:119]: 👌  thx, makes sense.

Comment (atigyi) [reply src/tests/utests/iotc_utest_memory_calloc.c:119]: SIZE_MAX is more than 18M TerraBytes... 🤦‍♂️

Comment (atigyi) [reply src/tests/utests/iotc_utest_memory_calloc.c:119]: so we can assume here it returns NULL pointer here... for years.

**Review by atigyi — APPROVED** [review]
LGTM + thx!! 👍
